### PR TITLE
Migration to avoid duplicate rows in code reviews

### PR DESCRIPTION
**Pull request by maureensturgeon**
Currently, we have several tables with unique indexes which aren't working properly and are resulting in duplicate rows. The `code_reviews` table is also susceptible to unwanted duplicate rows, it doesn't have any yet, but we'd like to take care of this issue now before the table gets any bigger. Here's the solution:
- add new virtual columns `open` and `active`
- replace `closed_at` and `deleted_at` in the unique index with these new columns

Before the deploy with this migration, I'll run the query `select count(*) as count_reviews from code_reviews where deleted_at is null and closed_at is null group by user_id, project_id having count_reviews > 1;` in production to ensure that there haven't been any duplicate rows added since this PR and to ensure the unique index can be created.

## Links
- jira ticket: [LP-2414](https://codedotorg.atlassian.net/browse/LP-2414)

## Testing
Ran migration locally and opened, closed code reviews

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/db/schema.rb [hunk 1]
```diff
@@ -320,7 +320,7 @@
     t.index ["code_review_request_id"], name: "index_code_review_notes_on_code_review_request_id"
   end
 
-  create_table "code_review_requests", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci", force: :cascade do |t|
+  create_table "code_review_requests", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci", force: :cascade do |t|
```
Comment (bethanyaconnor) [inline dashboard/db/schema.rb]: I don't think you want these utfmb4 changes

File: dashboard/db/schema.rb [hunk 2]
```diff
@@ -2205,17 +2207,6 @@
     t.index ["word", "definition"], name: "index_vocabularies_on_word_and_definition", type: :fulltext
   end
 
-  add_foreign_key "ap_school_codes", "schools"
```
Comment (bethanyaconnor) [inline dashboard/db/schema.rb]: Add back the foreign keys?

**Review by bethanyaconnor — APPROVED** [review]
This is super cool!

File: dashboard/db/schema.rb (hunk 1 above)
Comment (maureensturgeon) [reply dashboard/db/schema.rb]: Oops, yes thank you, good catch.

File: dashboard/db/schema.rb (hunk 2 above)
Comment (maureensturgeon) [reply dashboard/db/schema.rb]: Oh these too! Not sure how these are being generated

**Review by jamescodeorg — APPROVED** [review]
Thanks for taking care of this!

Were you able to test the index in the various cases by inserting / updating records locally?

**Comment by maureensturgeon** [comment]
> Thanks for taking care of this!
> 
> Were you able to test the index in the various cases by inserting / updating records locally?

I did run the migration and test that code review opening/closing still works. I could do more extensive testing if you think it's necessary.

**Comment by jamescodeorg** [comment]
> > Thanks for taking care of this!
> > Were you able to test the index in the various cases by inserting / updating records locally?
> 
> I did run the migration and test that code review opening/closing still works. I could do more extensive testing if you think it's necessary.

I think it would be good to try manually inserting / updating some rows in the database to simulate the important cases (some of which would be hard to trigger in the UI).  For example, insert rows with the same user_id, project_id, but the four different combination of null/not null for the closed_at and deleted_at columns.

**Comment by maureensturgeon** [comment]
> > > Thanks for taking care of this!
> > > Were you able to test the index in the various cases by inserting / updating records locally?
> > 
> > 
> > I did run the migration and test that code review opening/closing still works. I could do more extensive testing if you think it's necessary.
> 
> I think it would be good to try manually inserting / updating some rows in the database to simulate the important cases (some of which would be hard to trigger in the UI). For example, insert rows with the same user_id, project_id, but the four different combination of null/not null for the closed_at and deleted_at columns.

I was able to test some combinations and inserting or updating a row which would create a duplicate fails with the new index

**Pull request merged**